### ISSUE TITLE
[Regression test required] [$500] HIGH: Add `Avatar > About > Troubleshooting` section 

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#34082]
**Strategy:**
A billion users means a billion problems.  To diagnose and solve them at scale, we capture client-side logs and upload centrally.

**Problem:**
Some problems can't be diagnosed cleanly after the fact from the logs, for a few reasons:

- External contributors don't have access to production logs.
- Internal engineers do have production access, but they're often cumbersome to access.
- The logs are often ambiguous, such as if you have multiple devices connected and tabs open at the same time -- reverse engineering what happened for a particular client is confusing and time consuming.
- Observing logs in realtime is cumbersome or sometimes impossible, which slows down and complicates testing.
- Even if you have all the access you need, when talking with a user sometimes it's difficult to find the logs specific to them.

**Solution:**
Let's expand our ability to remotely diagnose production devices with better dev tools built into the app itself.  Specifically:

- Add a `Troubleshooting` section to the `Avatar > About` page, just above `Report a bug`
    - This will always be available, both on staging and production builds
- Move the following `Test preferences` settings into this new menu:
    - Use staging server
    - Force offline
    - Simulate failing network requests
    - Authentication status: Invalidate
    - Device credentials: Destroy
- Add `Reset and refresh`, which wipes Onyx clean except for the minimum needed to call OpenApp (so you don't need to reauthenticate)
- Add `View console` which opens a simple console:
    - This would first open up and initialize with all locally stored logs, so you can scroll back and see what happened
    - It would also append new logs in realtime
    - There would also be a `Save` button that just copies all available logs into a text file that is saved on the device
    - Add a `Share` button that brings up a `Search` selector and just posts the logs into the selected room (and then leaves the user navigated there, so they can talk about them).  I'm imagining this would be useful for chatting with a real world user and asking them to do something and send you the logs.
    - At the bottom is a command line that works similar to the Chrome console, where you can execute arbitrary JavaScript, with the results output to the console

This is blocking https://github.com/Expensify/App/issues/33040.
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c2a08b8b8ae90e21</li>
        <li>Upwork Job ID: 1744202378047897600</li>
        <li>Last Price Increase: 2024-01-15</li>
<li>Automatic offers: </li>
<ul>
<li>fedirjh | Reviewer | 28115102</li>
</ul></li>
    </ul>
</details>

**Comment by kmwamasali** [comment]
I would like to work on this feature

**Comment by shubham1206agra** [comment]
Can I directly work on this without a proposal?

**Comment by Tony-MK** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Some problems can't be diagnosed cleanly after the fact from the logs, for a few reasons

### What is the root cause of that problem?

The settings page is not optimal for diagnosing.

### What changes do you think we should make in order to solve the problem?


Add a new page in settings called `Test preferences`. 
Then move the necessary components to the new page.
Create a menu item on the settings menu for the page.

Then add another item in `Test preferences` page called `Reset and refresh`. This option will clear the onyx collection except for the auth data. Finally, it will refresh the page.

**Comment by quinthar** [comment]
Please submit proposals, especially for how you intend to do the console
feature. The more detail you provide, the greater likelihood we select your
proposal.

**Comment by samsonmobisa** [comment]
This is a thoughtful and comprehensive proposal that tackles the challenge that's encountered in diagnosing and solving issues at scale. I appreciate the clarity in the suggested changes, especially the addition of the Troubleshooting section and the integration of better dev tools within the app.

The proposed enhancements, such as the Console View, Reset and Refresh option, and the ability to Share Logs, would significantly improve the ability to remotely diagnose production devices. I particularly like the idea of having a command line for executing JavaScript, resembling the Chrome console.

I'm eager to contribute to the implementation of these features. Before diving in, is there any specific technical considerations or preferences regarding the implementation? Additionally, if there are any specific testing scenarios or edge cases to keep in mind, please provide guidance.

Looking forward to collaborating on this improvement

**Comment by Tony-MK** [comment]
@quinthar I have a couple of questions about the expected result.
1) Should the `View console` page be full screen due to the fact wide lines of code will be displayed and the console input component be wide enough to type in scripts?

2) Should we have a separate output log component to **differentiate** the output from the user-performed scripts? Another option is to color or tag it differently from the rest of the application's console log.

3) What action will the `Troubleshoot` button take?

**Comment by samsonmobisa** [comment]
Contributor details
Your Expensify account email: samsonmobisa62@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~018e78ebb343f52495

**Comment by samsonmobisa** [comment]
Have already submitted my details

**Comment by Piyush-Kumar-Ghosh** [comment]
**Proposal**

**Problem:**
Diagnosing and solving issues at scale is challenging due to limitations in accessing production logs, especially for external contributors. The logs are often cumbersome to access, ambiguous, and observing them in real-time can be complicated. This hinders effective testing and troubleshooting, impacting the reliability of the application.

**Root Cause:**
The current lack of robust dev tools within the app itself prevents efficient diagnosis of production devices, leading to difficulties in identifying and resolving issues.

**Solution:**
To address these challenges, the proposed solution involves expanding the app's ability to remotely diagnose production devices by integrating better dev tools. The key features of the proposed solution are:

1. **Troubleshooting Section:**
   - Add a Troubleshooting section to the Avatar > About page, just above Report a bug.
   - Available on both staging and production builds for accessibility.

2. **Test Preferences Settings:**
   - Move existing Test preferences settings (e.g., Use staging server, Force offline) into the Troubleshooting menu.

3. **Reset and Refresh:**
   - Introduce a Reset and refresh option that wipes Onyx clean except for the minimum needed to call OpenApp, facilitating a quick reset without reauthentication.

4. **View Console:**
   - Implement a simple console feature accessible from the Troubleshooting menu.
   - The console initializes with locally stored logs, allowing users to scroll back and review past events.
   - Real-time appending of new logs for immediate observation.
   - Save button to copy all available logs into a text file saved on the device.

5. **Share Logs:**
   - Add a Share button that brings up a Search selector, enabling users to post logs into a selected room.
   - Enhances communication with users for issue resolution.

6. **Command Line:**
   - Include a command line at the bottom, functioning similarly to the Chrome console.
   - Allows execution of arbitrary JavaScript for advanced debugging.

**Implementation Considerations:**
- Ensuring seamless integration with the existing app UI and user experience.
- Compatibility with different network conditions.
- Comprehensive testing scenarios to cover all use cases.
- Clear documentation for users and contributors.

I look forward to collaborating on the implementation of these features. If there are specific technical considerations or testing scenarios, please provide guidance. Your feedback and insights will be invaluable in enhancing the app's diagnostic capabilities.

**Comment by Piyush-Kumar-Ghosh** [comment]
Contributor details
Your Expensify account email: piyushkumarghosh1@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~017087736e94a4f9b8

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
HIGH: Add Avatar > About > Troubleshooting section

### What is the root cause of that problem?

NA - Feature request.

### What changes do you think we should make in order to solve the problem?

1. Creating a new Menu i.e TroubleShooting.

We already have `About` page in which we show multiple menu's under it, we should use `MenuItem` component to render the new `TroubleShoot` item.

Once user clicks on it, use `Navigation` to navigate to the troubleshoot page.

2. Moving `Test Preferences` to `TroubleShoot` page, since this `Troubleshoot` will be a new page, we need to move `MenuItem`'s from the `Preferences` page.

3. `Reset and Refresh`, we only need to wipe the onyx keys which doesn't require the user to be login again, this means we need to remove the data inside onyx that are not needed.

The majority of data we store are related to `reports`, `transactions`, `reportActions`, `policy`, `reportDraftComment`, `network`. 

To Simply say we should be having onyx data related to following keys `account`, `session`, `credentials`, `loginList`, `user`, after clearing the keys we should do `refresh` i.e `window.location.reload` in browser and some native code for native devices to restart the app. Maybe using `react-native-restart`.

4. `View console`
    
    1. Since console might contain a big chunk of lines i.e in real scenario people can be using the application for days where logs can be in 1000s of lines. We have logs stored in `Log.logLines` which we can use to render logs. 

    2. How, rendering well the logs are large list, which can block the low-end browsers to hang and doesn't give good UX. We should be using some virtualized list to render the logs, I prefer `react-virtual` for web and `react-native-fast-list` for native devices or we can simply use the `SectionList` that we're using to render the `reportActions`.

    3. Realtime logs rendering, well at the current code we cannot do with the current implementation, to resolve it we should be having some realtime i.e a variable that causes re-render with the code. 

    4. Saving, we should be generating a blob with type `text/plain` and download it using native tools, i.e for web we should be using `url.createObjectURL` and same implementation or `react-native-fs` to create the file.

    5. Sharing, when we click on `Share` we should snapshot the current logs (not new logs, because we're thinking about the current logs) and use the basic logic to share the logs.


### Implementation specifics / issues

Realtime rendering is the main issue here which will require some observable implementation.

Why? we're not saving `logLines` in the `react` specific code which doesn't let us use some react specific sections i.e `useState` or so. While figuring out the solution, I found the following package `https://jotai.org/docs/guides/react-native` which doesn't need any react specific code to create `observable` and can also be used with `react` that gives the ability to cause re-renders. So we should be creating a atom that stores the logs and should be exported and used in `App` codebase.

We have a limit of 10000 characters inside composer, but we don't know that `logLines` does contain the `10000` characters which will block the implementation of `Share` feature. Maybe we can wrap the logs in a file and share it with the selected chat. @quinthar you're view on this?



### What alternative solutions did you explore? (Optional)
NA

**Comment by samsonmobisa** [comment]
Hello Expensify Team

Your comprehensive plan demonstrates a deep understanding of the challenges at hand and offers a thoughtful and well-structured solution. I particularly appreciate the attention to detail in addressing each aspect of the problem. The proposed Troubleshooting section, along with the seamless integration of Test Preferences and the Reset and Refresh feature, reflects a strategic approach to streamline user support and enhance debugging processes.

The innovative use of observable atoms from the jotai package for real-time log rendering showcases a keen problem-solving mindset. Your consideration of potential issues, such as the character limit for sharing logs, reveals a thorough analysis of the implementation specifics.

This proposal not only tackles the immediate problem but also reflects a commitment to ensuring a robust, user-friendly application. I'm impressed by the holistic approach, covering everything from the user interface to the technical intricacies of real-time rendering.

I am ready and eager to collaborate on implementing these enhancements. Your feedback and guidance will be invaluable as we work together to create a more efficient and reliable workflow.

**Comment by quinthar** [comment]
What's the next step?  Who is doing it and when will it be done?

**Comment by Piyush-Kumar-Ghosh** [comment]
According to guidelines, All the proposals are reviewed by the Core Team from which one gets selected and then needs to do some follow up steps.

After that the contributor starts the work.

I haven't contributed to this project before so I would love to work on it.

**Comment by greg-schroeder** [comment]
Bump @fedirjh for proposal review

**Comment by samsonmobisa** [comment]
Contributor details
Your Expensify account email: samsonmobisa62@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~018e78ebb343f52495

**Comment by greg-schroeder** [comment]
bump @fedirjh

**Comment by quinthar** [comment]
Bump, what's the next step here?  I think @TMisiukiewicz has already volunteered.  Who are we waiting on?  What's the next step, and when will it happen?

**Comment by TMisiukiewicz** [comment]
As discussed with @muttmuure in #33040 , I want to take this over and start working on a solution. Can I get assigned to this issue?

**Comment by quinthar** [comment]
@TMisiukiewicz sounds great, go for it!

**Comment by TMisiukiewicz** [comment]
Quick update from my side:
- created `Troubleshooting` page and linked it with routing
- created "Reset and refresh" and "View console" menu items
- moved "Test preferences" section to "Troubleshooting" screen
- created basic implementation for "Reset and refresh", it works most of the times but I'm still figuring out what exact Onyx keys should be preserved from cleanining.

Do we have any design for the new page? If we want to follow the design of other pages, what i need to know is:
- icon to use with "Troubleshooting" menu item
- lottie illustration to use in the page header
- icon for "Reset and refresh" menu item
- icon for "View console" menu item
- design for console view (?)

**Comment by dubielzyk-expensify** [comment]
Unsure exactly what the console view looks like, but here's my take on how we could design the page. cc @Expensify/design 

![CleanShot 2024-01-24 at 17 27 54@2x](https://github.com/Expensify/App/assets/145416827/c0c5bb8b-5231-4d01-bda4-b7cdce9d769b)

**Comment by shawnborton** [comment]
I love that! I think that's a great illustration choice, and I actually think we have an animated version of it where the planet in the background rotates and the screen flashes: https://www.dropbox.com/scl/fi/k6n1k21839t4gs6m8s3x5/ED-workspaces-1-030323-kjs-4.gif?rlkey=f4iqy7paop13448l4dylrc92h&dl=0

Small nitpick, but I think the top page header would stay with a blue background even as content on the page scrolls. Check the existing examples in product to see how that works.

**Comment by TMisiukiewicz** [comment]
Thanks for the design! For those kind of illustrations we use Lottie in the app - is there any way to get it as a `.lottie` file?

> Unsure exactly what the console view looks like

I think it could be something similar to the console in the browser, just matching the mobile view and some other requirements mentioned above

https://github.com/Expensify/App/assets/13985840/9aedd9e1-579b-4111-bbdc-450486990aef

**Comment by shawnborton** [comment]
@TMisiukiewicz try this for the animation: https://www.dropbox.com/scl/fi/s1r6oty2q20fyydl148e6/ED-workspaces-1-030323-kjs-4.json?rlkey=f3hwuv9i5xcjlzuhynp0dwsrb&dl=0

**Comment by TMisiukiewicz** [comment]
Works nicely, thank you! 👍 
![image](https://github.com/Expensify/App/assets/13985840/17853f53-a0c0-444d-aadc-52cb6221566b)

**Comment by TMisiukiewicz** [comment]
I also need Spanish translations for the following phrases:
- Troubleshoot
- Reset and refresh
- View console
- Use the tools below to help troubleshoot the Expensify experience. If you encounter any issues, please submit a bug

For wiping the Onyx state, I preserve the Onyx keys that seem to be static or relevant to user account: `account`, `currency_list`, `device_id`, `isFirstTimeNewExpensifyUser`, `isLoadingApp`, `isSidebarLoaded`, `loginList`, `nvp_priorityMode`, `preferredLocale`, `preferredTheme`, `session`, `tryFocusMode`, `user`, `userWallet`. Can someone who is more experienced with Onyx take a look and verify if I can safely remove something from this list, or if there are any missing keys I need to add?

Can we split work on this issue into 2 PRs? First would introduce the Troubleshooting page and "Reset and refresh" feature and second one would contain the console. The console feature will probably take some time, so this way we could ship the "Reset and refresh" feature faster

**Comment by muttmuure** [comment]
cc @cead22 @pecanoro

**Comment by pecanoro** [comment]
@muttmuure Please, follow this if you need some translations! https://stackoverflowteams.com/c/expensify/questions/10316/10317

**Comment by muttmuure** [comment]
Oops, sorry!

**Comment by dubielzyk-expensify** [comment]
>Small nitpick, but I think the top page header would stay with a blue background even as content on the page scrolls. Check the existing examples in product to see how that works.

My bad on the explanation. It was actually a variation with and without an illustration. Not meant to be a scrolled version 😅 But yeah we should do what you said

**Comment by shawnborton** [comment]
Ah got it, all good and sounds good to me!

**Comment by shawnborton** [comment]
For this screenshot [here](https://github.com/Expensify/App/issues/34082#issuecomment-1907718669) though, looks like it doesn't quite match your mockup @dubielzyk-expensify - I think you are using blue600 or blue700 as the BG?

**Comment by TMisiukiewicz** [comment]
@shawnborton Good catch, thanks!

Additionally, what icon should I use for Troubleshoot menu item? The one you see on the screenshot below is a placeholder for now
<img width="368" alt="image" src="https://github.com/Expensify/App/assets/13985840/32f51768-306e-49e3-917e-0b35b9d604f3">

**Comment by shawnborton** [comment]
Hmm good question. Curious what the @Expensify/design thinks but perhaps a lightbulb or the "!" (exclamation) icon could work?

**Comment by dannymcclain** [comment]
I like the lightbulb or the "!" icon. I lean toward the "!" but I want to make sure it doesn't feel _too_ alarming haha.

![CleanShot 2024-01-25 at 08 26 14@2x](https://github.com/Expensify/App/assets/7516624/bc599f23-8318-4d18-b0f4-2bc3147df725)

**Comment by shawnborton** [comment]
I think both of those icons are in the repo already, so perhaps @TMisiukiewicz can try each of those quickly and we can see which one we prefer? I could go either way... I do like that the lightbulb has a more positive connotation to it though!

**Comment by TMisiukiewicz** [comment]
@shawnborton @dannymcclain I couldn't find the lightbulb in the repo, here's the example view with Exlamation icon
![image](https://github.com/Expensify/App/assets/13985840/128ce6da-9cb0-435f-82a2-dcd4d888f7d6)

**Comment by dannymcclain** [comment]
I don't hate that, but I'm still curious to see the lightbulb icon.

Weirdly, I found the lightbulb here: `assets/images/emojiCategoryIcons/light-bulb.svg`
Not sure why it's not in the main images directory with all the other icons as well though. Does that help at all?

**Comment by TMisiukiewicz** [comment]
Thanks for finding it, here's a version with a lightbulb:
![image](https://github.com/Expensify/App/assets/13985840/439ae547-c24e-4062-9db7-0a4112b6a4ba)

**Comment by shawnborton** [comment]
Big fan of that! Let's see what @dannymcclain and @dubielzyk-expensify think though!

**Comment by dannymcclain** [comment]
Yeah I like the lightbulb way better! Much less anxiety producing haha.

**Comment by dubielzyk-expensify** [comment]
Yeah. Lightbulb is cute. Dig it!

**Comment by TMisiukiewicz** [comment]
Do we need any message to display after pressing "Reset and refresh"? This might be confirmation modal (e.g. "Are you sure you want to wipe the data?"), or alert/toast saying the data was wiped. If so, what message should be there?

![image](https://github.com/Expensify/App/assets/13985840/1319b066-c8c6-4af2-930d-ba2a74ef5939)

**Comment by TMisiukiewicz** [comment]
Opened up a Draft PR! While going through the checklist, noticed new copies have to be approved by marketing team. Can someone take a look on the copies suggested in the design above?

- Reset and refresh
- View console
- Use the tools below to help troubleshoot the Expensify experience. If you encounter any issues, please submit a bug

Once they are approved, I'll ask for Spanish translations

**Comment by TMisiukiewicz** [comment]
I'm also about to start working on a Console, it's not blocking me in any way for now but is there any update regarding the design for it?

**Comment by shawnborton** [comment]
> I'm also about to start working on a Console, it's not blocking me in any way for now but is there any update regarding the design for it?

What does this mean exactly?

**Comment by TMisiukiewicz** [comment]
@shawnborton wiping the data basically means clearing all the Onyx data except those keys that are needed to run the app as a freshly logged-in user. There are some keys that could affect the user account, e.g. any draft messages would be gone.

You can see the currently implemented behavior on both web and mobile here:

https://github.com/Expensify/App/assets/13985840/f8996dc9-4d05-414a-9f55-e3b806a18de8

https://github.com/Expensify/App/assets/13985840/f02f6655-6a8b-44c1-8863-96cfc0e46748

As you can see, currently there is no full page spinner/splash screen after pressing it. Should we approach it differently?



> >I'm also about to start working on a Console, it's not blocking me in any way for now but is there any update regarding the design for it?
>
> What does this mean exactly?

I'm talking about "View console" feature mentioned in the [first post of this issue](https://github.com/Expensify/App/issues/34082#issue-2069570061). I assume it would look and work pretty similar to the console known from web browsers e.g. Chrome. Would be great if we could figure out how to implement this view in the app

**Comment by shawnborton** [comment]
> wiping the data basically means clearing all the Onyx data except those keys that are needed to run the app as a freshly logged-in user. There are some keys that could affect the user account, e.g. any draft messages would be gone.

Got it, thanks for clarifying. I could see where a small alert dialog would work in that case, which exists as an existing pattern (for instance, try to delete an outstanding money request and you will see the confirmation modal). Curious if @quinthar agrees though!

> You can see the currently implemented behavior on both web and mobile here:

Got it. I think this is fine, especially if we have the confirmation modal. 

> I'm talking about "View console" feature mentioned in the https://github.com/Expensify/App/issues/34082#issue-2069570061. I assume it would look and work pretty similar to the console known from web browsers e.g. Chrome. Would be great if we could figure out how to implement this view in the app

I'm not entirely sure what the intention is behind this feature, so would love for @quinthar to chime in on this one too. Are you thinking we would display our own custom view of a console, or just launch the existing browser console, or something else?

**Comment by TMisiukiewicz** [comment]
I was able to create a basic implementation of in-app console that executes the arbitrary code, here is the demo:

https://github.com/Expensify/App/assets/13985840/2226aa3a-3798-43f3-90e4-51b8364eefe0

Next thing coming is connecting it with existing logs. Still waiting for any guidelines regarding the design.

@muttmuure could you help me get an approval for new copies listed in [this comment](https://github.com/Expensify/App/issues/34082#issuecomment-1914524197)? This would speed up opening the first PR with Troubleshoot page 🙏

**Comment by LLPeckham** [comment]
Hiya! Added a few clarifying questions below for the first two bullets just so we can possibly make those more specific. 

cc @muttmuure @TMisiukiewicz 

- **Reset and refresh** -- For this one, @TMisiukiewicz, I had the same questions as @shawnborton when reviewing this copy. can you tell me what exactly is being reset when someone clicks this? I'm thinking we should a tiny bit more detail so someone knows what exactly they're doing when they click this. Possibly something like "Reset Onyx data and refresh" Or something like this?
- **View console** -- For this one, what is the console for? Can we add a descriptor so someone knows what they'd be opening exactly when clicking on this? 
- I think the 3rd bullet is fine as is but we should add a period at the end of "please submit a bug."

**Comment by muttmuure** [comment]
My two cents:

- Reset and refresh _stashed data_ 

- View _debug_ console

Are more specific and help indicate what these new features do

**Comment by LLPeckham** [comment]
I'd be down for those, @muttmuure. I think just adding a little more context helps in this case. So we'd have: 

Use the tools below to help troubleshoot the Expensify experience. If you encounter any issues, please submit a bug.
- Reset and refresh stashed data
- View debug console

**Comment by TMisiukiewicz** [comment]
> I'm thinking we should a tiny bit more detail so someone knows what exactly they're doing when they click this.

Agree on that one, however we might give a detailed description in a confirmation modal after pressing it

> Possibly something like "Reset Onyx data and refresh" Or something like this?

Users probably don't know what Onyx is, so I think "Reset and refresh stashed data" proposed by @muttmuure should be fine and we can think about a little bit detailed description in a confirmation modal as I mentioned above

**Comment by LLPeckham** [comment]
Ok cool, so let's move forward with [this](https://github.com/Expensify/App/issues/34082#issuecomment-1919081261) for the copy if you agree, @TMisiukiewicz?

**Comment by TMisiukiewicz** [comment]
sounds great to me @LLPeckham 👍 Thank you!

**Comment by TMisiukiewicz** [comment]
Since copies are approved, I sent the draft PR #35306 for our internal review, I'll open it once I have an approval from CK side.

Today I was also moving forward with the console. I've already integrated existing logs with it and added functionality of saving it on a device. Tomorrow I'll be working on displaying real-time logs as well

https://github.com/Expensify/App/assets/13985840/540e555e-4a70-494c-a986-215ce34459d6

**Comment by shawnborton** [comment]
We can help provide better designs for this page:
![CleanShot 2024-01-31 at 10 18 35@2x](https://github.com/Expensify/App/assets/2319350/969803b8-d780-470f-b9ef-b0dd8325cb11)

But to get you started, we should open up a new page in the RHP dedicated to this entire thing. So the page/view would be called Console.

**Comment by shawnborton** [comment]
@dubielzyk-expensify given that you worked on the [original design](https://github.com/Expensify/App/issues/34082#issuecomment-1907547333) for this, let us know if you want to take a stab at the console page design too. 

Looks like the brief for that one is:

> Add View console which opens a simple console:
> 
> - This would first open up and initialize with all locally stored logs, so you can scroll back and see what happened
> - It would also append new logs in realtime
> - There would also be a Save button that just copies all available logs into a text file that is saved on the device
> - Add a Share button that brings up a Search selector and just posts the logs into the selected room (and then leaves the user navigated there, so they can talk about them). I'm imagining this would be useful for chatting with a real world user and asking them to do something and send you the logs.
> - At the bottom is a command line that works similar to the Chrome console, where you can execute arbitrary JavaScript, with the results output to the console

**Comment by shawnborton** [comment]
Looks great to me! Nice job on this.

**Comment by TMisiukiewicz** [comment]
The design looks great! 👍 I already started slowly implementing it today.

I have more technical problem now. Today I started working on adding real-time logs, that's why i thought switching my implementation to Onyx would be good. My idea was simply merging new logs when they appear in the logger, however that generates a problem of infinite loop, since we are getting `merge() called...` log endlessly in this scenario. So a few questions from my side:
- what logs should appear in the console? Basically each log that is currently printed in browser console in development mode?
- if Onyx logs are required, is there a built-in way to determine what logs should be ignored so we could avoid `merge() called for key logs` messages?

**Comment by dubielzyk-expensify** [comment]
100% with you on Mono font. I thought about it but was unsure if we even had access to it in the app. So glad you brought this up <3 Updated mocks with mono font:
![CleanShot 2024-02-02 at 08 18 28@2x](https://github.com/Expensify/App/assets/145416827/507da6a0-7e19-4e7c-ba4c-1fe457d4c636)

**Comment by shawnborton** [comment]
> I thought about it but was unsure if we even had access to it in the app.

We totally have access to it already! Just try using ticks for inline code or three ticks for a code block!

**Comment by TMisiukiewicz** [comment]
Small update, I connected logs with Onyx and created a list of logs that should be ignored, so I can avoid infinite loop when merging new items. Now items are displayed in real-time when console is open, however now it does not display previously collected logs, so I have to figure out how to correctly collect them in the background. 

I also started working on a design implementation:

<img width="352" alt="image" src="https://github.com/Expensify/App/assets/13985840/1e136922-a70f-4f76-b833-143117454a10">

**Comment by shawnborton** [comment]
Looking good so far. For the buttons, why isn't the button text center-aligned?

**Comment by TMisiukiewicz** [comment]
Today I connected logs with Onyx and made it possible to display historical logs together with real-time logs. I feel like storing all the Onyx logs might be inefficient at some point, after some tests my file to download already had 100mbs. I think it'd be helpful to decide what exact logs should be stored there, eventually we should think of building some kind of mechanism that would flush old logs from the console. cc @quinthar 

Here's the demo showing historical logs, adding real-time logs and executing commands

https://github.com/Expensify/App/assets/13985840/9d82ab13-01e5-471e-9a46-bed7e3a0bf44

> Looking good so far. For the buttons, why isn't the button text center-aligned?

@shawnborton I was looking at the code responsible for displaying icon and text in the button and turns out when the left or right icon is displayed, the text is automatically aligned to the left 🤔 should it work different way?

**Comment by shawnborton** [comment]
It should always be center-aligned as far as I know... cc @Expensify/design for additional thoughts though.

**Comment by dannymcclain** [comment]
> It should always be center-aligned as far as I know... cc @Expensify/design for additional thoughts though.

Definitely. Those buttons should look like [Jon's mock](https://github.com/Expensify/App/issues/34082#issuecomment-1922364871).

**Comment by dubielzyk-expensify** [comment]
Yep, what Shawn and Danny said. Also worth mentioning that the logs area is the only area that should scale:
![CleanShot 2024-02-06 at 09 40 29@2x](https://github.com/Expensify/App/assets/145416827/63916963-03dc-44c9-833b-a5c0fce3ba6d)

**Comment by greg-schroeder** [comment]
Okay so according to https://github.com/Expensify/App/issues/34082#issuecomment-1923783535 this needs to be Internal instead, so I've adjusted it accordingly.

@techievivek you were assigned by the C+ review - do you want to take this, or should we make it a `Hot Pick` and call it out in the update? cc @quinthar

**Comment by quinthar** [comment]
Wow, this is all amazing!  Just to confirm, the logs aren't saved to Onyx until you open this troubleshooting dialog, right?  We don't want to do anything that will affect the 99.99% of users who _won't_ open this dialog.    Where are the logs normally stored, prior to this change?

**Comment by TMisiukiewicz** [comment]
@quinthar right now it's all stored in the Onyx from the beginning since we want to see historical logs or actions that happened when the app is backgrounded. I agree that is not the optimal solution as it would affect all the users. We can easily change it to collect logs only if user opens up a troubleshooting page, however once it is done, there would be no way to disable it other than logging out/flushing Onyx data, which may lead to uncontrolled amounts of data stored in Onyx . As an alternative, I'd suggest making logs collector as an opt-in, so they would be stored in Onyx only if the user specifically enables it with toggle (pretty much the same way the "Developer options" on Android works). This way those who want to track their actions might enable or disable it whenever they need. Disabling would automatically clear all the logs stored in Onyx. If the toggle is off, "View debug console" should not be visible. WDYT?

**Comment by shawnborton** [comment]
> button styles updated ✅ once we have it merged this style will apply globally

Looks great, thanks!

**Comment by TMisiukiewicz** [comment]
Today's update:
- applied styles for buttons to center the label and icon
- applied styles to scale the console window ([comment](https://github.com/Expensify/App/issues/34082#issuecomment-1928512986))
- created new screen to select a chat to share the log

Tomorrow I have a plan to implement sending log for selected report

https://github.com/Expensify/App/assets/13985840/621f359b-7dd0-4c3d-9be2-18922e292430

**Comment by TMisiukiewicz** [comment]
hi @quinthar  today I’ve been working on sharing the logs. I was investigating how logged data grows on a mobile and how could we effectively collect it. So right now I have ~2300 logs which is around 2,1mb. My first idea was to send it as formatted message, however composer accepts maximum 10000 characters. So sending this as a txt file seems more efficient, as it gives a possibility to send a file up to 24mbs. We could automatically remove logs older than e.g. 3 days to keep it as lean as possible and avoid building files that exceeds this limit. What do you think about this solution? 

Additionally, do we want to display all logs that are going through logger, or do we want to collect only some of them? Second solution would definitely have a positive impact on a size of the file

I also posted an idea to make collecting logs an opt-in [here](https://github.com/Expensify/App/issues/34082#issuecomment-1929093453). Wondering what are your thoughts about this, since as you mentioned, we don’t want to affect 99% of the users who won’t open the debug console.

**Comment by techievivek** [comment]
@greg-schroeder 

> I guess I see @TMisiukiewicz working on part of it, so is the internal engineer working on a separate PR or just doing the review stages alongside C+? Sorry, just a bit confusing here

We don't need any backend changes for this, we will be just helping doing the reviews along with C+

**Comment by techievivek** [comment]
@TMisiukiewicz 

> right now it's all stored in the Onyx from the beginning since we want to see historical logs or actions that happened when the app is backgrounded. I agree that is not the optimal solution as it would affect all the users. We can easily change it to collect logs only if user opens up a troubleshooting page, however once it is done, there would be no way to disable it other than logging out/flushing Onyx data, which may lead to uncontrolled amounts of data stored in Onyx 

Can we not have a job in the background that could wipe the logs at regular intervals?

**Comment by techievivek** [comment]
> As an alternative, I'd suggest making logs collector as an opt-in, so they would be stored in Onyx only if the user specifically enables it with toggle 

In the worst case, we would still face the same issue where the onyx will be flooded with the log data, so I think we will need to look into a solution that is scalable for everyone.

**Comment by TMisiukiewicz** [comment]
@techievivek 
> Can we not have a job in the background that could wipe the logs at regular intervals?

Discussed it lately [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1707324615099869?thread_ts=1703982689.203629&cid=C01GTK53T8Q), however it's hard to decide what interval should we take here. My main concern is: the longer interval, the bigger possibility we could excess the 24 mb limit of the attachment file. 

Recently I discovered logs are being sent to the server (via Logger from expensify-common package) and I am wondering if we could access those logs so we could avoid keeping the historic data on the device.

**Comment by TMisiukiewicz** [comment]
Today I was able to implement sharing txt file with users on iOS:
 
https://github.com/Expensify/App/assets/13985840/6ee12393-46be-4132-a2c0-0b9e62e99b61

I still have to do some debugging on Android, as i am getting "Unexpected error" when trying to upload a file. Still need to work on Web/desktop implementation as I cannot use a filesystem there, trying to figure out how to achieve the same result with Blob

I have also added a flag in Onyx that determines if the logs should be collected. By default it's false, and it's set to true when user opens up the Debug console for the first time. Happy to discuss more improvements in this area, as we have to make sure the file does not exceed 24mbs.

**Comment by TMisiukiewicz** [comment]
@shawnborton @dubielzyk-expensify could you help me with the design for [this comment](https://github.com/Expensify/App/issues/34082#issuecomment-1929093453)? Discussed it with @quinthar [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1707451360304539?thread_ts=1703982689.203629&cid=C01GTK53T8Q)

**Comment by TMisiukiewicz** [comment]
While waiting for a proper design, I already implemented the switch to build a proof-of-concept how it could possibly work, here are the results containing:
- enabling client side logging
- saving file on the device
- sharing the logs as an attachment in a report


https://github.com/Expensify/App/assets/13985840/bf9fc5e7-ec98-4c9e-9b73-8f78ec17bf34

https://github.com/Expensify/App/assets/13985840/3efa5a2e-0d56-4545-96ae-606bd9f6395f

Sharing logs now works cross-platform, so we have full-working feature. On Monday I'll verify if zipping the `.txt` would help reduce the filesize (cc @muttmuure), and I'll do some cleanup on the branch. Still waiting for the proper design for switch and decisions regarding flushing the old logs. Additionally the console feature is blocked by #35306. I hope we'll figure this out quickly and we'll be able to create a draft PR on Monday 🤞

**Comment by muttmuure** [comment]
Wow this looks so great!

**Comment by muttmuure** [comment]
I am really looking forward to using this

**Comment by shawnborton** [comment]
> While waiting for a proper design

Can you break down exactly what you need here? I looked at your most recent POC and it looks really great!

**Comment by dubielzyk-expensify** [comment]
Agree with @shawnborton . What you have looks awesome. Happy to provide anything you need though

**Comment by techievivek** [comment]
@fedirjh 

> cc @quinthar What is the expected behavior for this feature in offline mode? Wiping the data may result in the app becoming unusable in offline mode. Should we consider enabling this option exclusively for online users, or is logging out the user in offline mode a more suitable approach?

Given that the app will become unusable, I think we can enable this option only for online users. Otherwise, if they are on a bad connection, we might end up making the App unusable for them.

**Comment by TMisiukiewicz** [comment]
@shawnborton @dubielzyk-expensify thanks, I'm gonna follow the @quinthar's suggestions from [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1707514482114469?thread_ts=1703982689.203629&cid=C01GTK53T8Q), no need for additional work from your side  👍  

@techievivek 
> I think we can benchmark it and agree on an interval that should be safe in the worst scenario. I don't think that would be something difficult, no?

I think it's difficult to benchmark it since logs may vary depending on an actions performed by user and the amount of data on user account (e.g. merge `personalDetails` log). Maybe let's just check the size of the blob when pressing "Share log",   and if it exceeds 24mb, just cut e.g. 1000 oldest logs and then move to another screen?

> Yes, we send the logs to the backend, which then gets stored for ~30 days in Elasticsearch, and we then look at them in Logsearch or Kibana. We can create a command that queries and returns logs, but we need a wider discussion for this. Let me know if you would want me to bring up a discussion on this.

Discussed potential sharing the logs with @muttmuure [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1707379892647659?thread_ts=1703982689.203629&cid=C01GTK53T8Q). I think we could stay with the currently implemented solution that stores logs on the client side

**Comment by TMisiukiewicz** [comment]
> @fedirjh
> 
> > cc @quinthar What is the expected behavior for this feature in offline mode? Wiping the data may result in the app becoming unusable in offline mode. Should we consider enabling this option exclusively for online users, or is logging out the user in offline mode a more suitable approach?
> 
> Given that the app will become unusable, I think we can enable this option only for online users. Otherwise, if they are on a bad connection, we might end up making the App unusable for them.

Can we discuss the detailed implementation for this? Should the button just be disabled when in offline mode? Or maybe should it display an alert with a message?

**Comment by TMisiukiewicz** [comment]
Opened up draft PR #36337, once we have #35306 merged and implement flushing old logs, it should be ready for review

**Comment by dannymcclain** [comment]
> Displaying an alert seems friendlier.

I agree! And that follows our standard form practices as well. (We try not to disable buttons as much as possible and instead opt to throw errors.)

**Comment by TMisiukiewicz** [comment]
Both Troubleshooting page and Debug console page are merged now 🎉 However they are currently displayed in an RHP, wondering if they should follow the new design and be displayed in full screen navigator? Could we get an input from someone from design team?
cc @fedirjh @techievivek

**Comment by techievivek** [comment]
CC @shawnborton @dannymcclain for your thoughts on [above](https://github.com/Expensify/App/issues/34082#issuecomment-1945622864) comment. 🙏

**Comment by LLPeckham** [comment]
Heya! Seeing I'm assigned here but not seeing any tags or reasons of why - was I accidentally added to this GH? @techievivek @muttmuure

**Comment by techievivek** [comment]
@LLPeckham Looks like you helped with some copy https://github.com/Expensify/App/issues/34082#issuecomment-1919076811

**Comment by LLPeckham** [comment]
Ahh yes. How quickly I forget. Thanks 😂

**Comment by shawnborton** [comment]
Hmm good question. I think technically Troubleshoot should only get the full card treatment if we make it its own standalone section in the LHN and pull it out of About. I don't think that's a bad idea... we might want to put it just below About in the Settings LHN? cc @muttmuure @Expensify/design for thoughts on that. But I think this one could quite easily be converted to that wider card style that other sections are using.

**Comment by dannymcclain** [comment]
> I think technically Troubleshoot should only get the full card treatment if we make it its own standalone section in the LHN and pull it out of About. I don't think that's a bad idea... we might want to put it just below About in the Settings LHN? 

I honestly don't feel super strongly either way. I agree that it could easily become it's own standalone page (and I like your suggestion of putting it underneath the `About` nav item).

**Comment by marcochavezf** [comment]
Hi @TMisiukiewicz! We got an alert in the backend that we're receiving a massive message log from NewDot, and it seems it's coming from this [PR](https://github.com/Expensify/App/pull/36337) that was deployed recently. This is an example of the message:
 
```
LogAPI - Massive log message given ~~ 0: '[info] [Onyx] merge() called for key: logs properties: 1708058686372,1708058686377,1708058696327,1708058730961,1708058732236,1708058732237,1708058734681,1708058734923,1708058734924,1708058734925,1708058734926,1708058734934,1708058734983,1708058734984,1708058734985,1708058734986,1708058735023,1708058735064,1708058735083,1708058735101,1708058735102,1708058735106,1708058735138,1708058735153,1708058735167,1708058735302,1708058735331,1708058735446,1708058735453,1708058735459,1708058735465,1708058735572,1708058735586,1708058735588,1708058735771,1708058735787,1708058735859,1708058735877,1708058735886,1708058735887,1708058735992,1708058735993,1708058736070,1708058736119,1708058736120,1708058737117,1708058737118,1708058737128,1708058737139,1708058741789,1708058741809,1708058741810,1708058741811,1708058741812,1708058741966,1708058741967,1708058741976,1708058742114,1708058748941,1708058749061,1708058749062,1708058749070,1708058749209,1708058749218,1708058749219,1708058749233,1708058749235,1708058749272,1708058749573,170'
```

We are already trimming the message in the backend using this logic:

```
if (strlen($log['message']) > 1024 * 1024) {
    self::alert('Massive log message given', substr($log['message'], 0, 1024));
} 
```

And I was wondering if we do the same in the frontend also to avoid sending large amounts of data for the logs, thoughts?

**Comment by TMisiukiewicz** [comment]
Hi @marcochavezf! That's weird because there is a condition to not attach logs with pattern `[Onyx] merge() called for key: logs` because otherwise it would create an infinite loop and end up with tons of huge log messages (since new log is attached to `logs`, which triggers logging a message that merge was called for key logs... and so on). Is this happening frequently? My only idea for now is someone did some changes in the codebase locally and removed this condition

**Comment by marcochavezf** [comment]
Oh interesting, I happened several times yesterday, and according to the logs it wasn't coming from a local environment. I will report again if the alert pops up again, thanks!

**Comment by greg-schroeder** [comment]
Reviewing

**Comment by greg-schroeder** [comment]
Paid!

@fedirjh can you create a regression test for this new feature? Thanks!

**Comment by techievivek** [comment]
Nice, regression tests look good to me. 👍 CC @greg-schroeder

**Comment by greg-schroeder** [comment]
LGTM! Filing!
